### PR TITLE
fix: creating duplicate params via renaming vars

### LINKED ISSUE #3767: Renaming vars so that you get duplicate procedure vars breaks mutator (sometimes)
**Describe the bug**

If you rename a variable via a variable block, it can cause duplicate variables in a procedure.

**To Reproduce**

1. Create a mutator with two variables.
2. Grab a variable block for one of the variables.
3. Rename the variable to match the other variable in the mutator.
4. Observe how there are duplicate vars in the procedure. (bad)

The next bit does not always happen, depending on which var you rename:
1. Open the mutator.
6. Observe how you have two vars.
7. Close the mutator and reopen.
8. Observe how you get errors in the console. 

**Expected behavior**

There are two parts:
1) Duplicate vars: No clue how this should be handled.
2) Errors: Errors are bad, definitely do not want errors.

**Screenshots**

Errors:
DuplicateVars

No Errors:
DuplicateVars2

**Desktop (please complete the following information):**
 - OS: Ubunto 18.04
 - Browser Chrome
 - Version 80

**Stack Traces**

```
procedures.js:110 Uncaught TypeError: Cannot read property 'name' of null
    at Blockly.BlockSvg.mutationToDom (procedures.js:110)
    at Object.Blockly.Procedures.mutateCallers (procedures.js:374)
    at Blockly.BlockSvg.compose (procedures.js:227)
    at Blockly.Mutator.workspaceChanged_ (mutator.js:377)
    at Blockly.WorkspaceSvg.Blockly.Workspace.fireChangeListener (workspace.js:657)
    at Blockly.Events.fireNow_ (events.js:192)
mutationToDom @ procedures.js:110
Blockly.Procedures.mutateCallers @ procedures.js:374
compose @ procedures.js:227
Blockly.Mutator.workspaceChanged_ @ mutator.js:377
Blockly.Workspace.fireChangeListener @ workspace.js:657
Blockly.Events.fireNow_ @ events.js:192
setTimeout (async)
Blockly.Events.fire @ events.js:174
Blockly.Mutator.setVisible @ mutator.js:274
Blockly.Icon.iconClick_ @ icon.js:124
Blockly.Mutator.iconClick_ @ mutator.js:124
wrapFunc @ blockly.js:462
variables.js:51 Uncaught TypeError: Cannot read property 'getId' of null
    at Object.Blockly.Variables.allUsedVarModels (variables.js:51)
    at Object.Blockly.Xml.workspaceToDom (xml.js:38)
    at exportWorkspace (my_playground.html:438)
    at autoUpdater (my_playground.html:561)
    at Blockly.WorkspaceSvg.Blockly.Workspace.fireChangeListener (workspace.js:657)
    at Blockly.Events.fireNow_ (events.js:192)
Blockly.Variables.allUsedVarModels @ variables.js:51
Blockly.Xml.workspaceToDom @ xml.js:38
exportWorkspace @ my_playground.html:438
autoUpdater @ my_playground.html:561
Blockly.Workspace.fireChangeListener @ workspace.js:657
Blockly.Events.fireNow_ @ events.js:192
setTimeout (async)
Blockly.Events.fire @ events.js:174
Blockly.Mutator.setVisible @ mutator.js:274
Blockly.Icon.iconClick_ @ icon.js:124
Blockly.Mutator.iconClick_ @ mutator.js:124
wrapFunc @ blockly.js:462
procedures.js:110 Uncaught TypeError: Cannot read property 'name' of null
    at Blockly.BlockSvg.mutationToDom (procedures.js:110)
    at Object.Blockly.Procedures.mutateCallers (procedures.js:374)
    at Blockly.BlockSvg.decompose (procedures.js:202)
    at Blockly.Mutator.setVisible (mutator.js:292)
    at Blockly.Mutator.Blockly.Icon.iconClick_ (icon.js:124)
    at Blockly.Mutator.iconClick_ (mutator.js:124)
    at SVGGElement.wrapFunc (blockly.js:462)
mutationToDom @ procedures.js:110
Blockly.Procedures.mutateCallers @ procedures.js:374
decompose @ procedures.js:202
Blockly.Mutator.setVisible @ mutator.js:292
Blockly.Icon.iconClick_ @ icon.js:124
Blockly.Mutator.iconClick_ @ mutator.js:124
wrapFunc @ blockly.js:462
variables.js:51 Uncaught TypeError: Cannot read property 'getId' of null
    at Object.Blockly.Variables.allUsedVarModels (variables.js:51)
    at Object.Blockly.Xml.workspaceToDom (xml.js:38)
    at exportWorkspace (my_playground.html:438)
    at autoUpdater (my_playground.html:561)
    at Blockly.WorkspaceSvg.Blockly.Workspace.fireChangeListener (workspace.js:657)
    at Blockly.Events.fireNow_ (events.js:192)
```

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by BeksOmega**
## The basics

- [X] I branched from develop
- [X] My pull request is against develop
- [X] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)
- [X] I ran `npm run format` and `npm run lint`

### Resolves

Fixes #3767 

### Proposed Changes

Adds a step to `renameVariable` that detects if a variable rename will create a parameter conflict.

Detects conflicts in legacy procedure blocks by walking the workspace, and conflicts in procedure models by checking if the parameter wraps a variable model.

Alternative implementations discussed here: https://github.com/google/blockly/issues/3767#issuecomment-1450422109

### Reason for Changes

Squash de buggies

### Test Coverage

Added tests for `nameUsedWithConflictingParam` for both legacy procedure blocks and procedure models.

### Documentation

N/A

### Additional Information

Best reviewed **commit-wise**, especially because I reorganize some code to return-early instead of deeply nesting.

### CODE REVIEW

File: core/interfaces/i_variable_holder.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import type {VariableModel} from '../variable_model.js';
+
+
+/** Interface for a data model that holds a variable model. */
+export interface IVariableHolder {
```
Comment (maribethb) [inline core/interfaces/i_variable_holder.ts:11]: I don't love this name. Also, this is only going to used in conjunction with `IParameterModel` right? How about

```
export interface IVariableBackedParameterModel extends IParameterModel {}
```

??

Then in the documentation and/or the plugin we can tell people to implement this one if the parameters will be backed by variables, or the base one if not.

File: core/variables.ts [hunk 2]
```diff
@@ -385,6 +397,49 @@ export function nameUsedWithAnyType(
...
+ *
+ * @param name The name to check for conflicts.
+ * @param workspace The workspace to search for conflicting parameters.
+ */
+function nameUsedWithConflictingParam(
+    oldName: string, newName: string, workspace: Workspace): string|null {
+  oldName = oldName.toLowerCase();
+  newName = newName.toLowerCase();
+
+  const blocks = workspace.getAllBlocks(false);
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply core/variables.ts:416]: I think checking if there's data in the procedure map should be sufficient!

**Review by maribethb — APPROVED** [review]
Please fix lint warning, then lgtm

**Comment by BeksOmega** [comment]
Fixed lint, and set it to error since we want to enforce it! Please tell me if you want me to revert that change.

File: msg/messages.js [hunk 3]
```diff
@@ -166,6 +166,9 @@ Blockly.Msg.VARIABLE_ALREADY_EXISTS = 'A variable named "%1" already exists.';
 /** @type {string} */
 /// alert - Tells the user that the name they entered is already in use for another type.
 Blockly.Msg.VARIABLE_ALREADY_EXISTS_FOR_ANOTHER_TYPE = 'A variable named "%1" already exists for another type: "%2".';
+/** @type {string} */
+/// alert - Tells the user that the name they entered is already in use as a parameter to a procedure, that the variable they are renaming also exists on. Renaming would create two parameters with the same name, which is not allowed.
+Blockly.Msg.VARIABLE_ALREADY_EXISTS_FOR_A_PARAMETER = 'A variable named "%1" already exists as a parameter in the procedure "%2".';
```
Comment (cpcallen) [inline msg/messages.js:171]: Looks like you did not `npm run messages` (formerly …`generate:langfiles`), so this message is not included in any translation (including English).

**Pull request merged**